presto-common/src/main/java/com/facebook/presto/common/plan/PlanCanonicalizationStrategy.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common.plan;
+
+public enum PlanCanonicalizationStrategy
+{
+    /**
+     * DEFAULT strategy is used to canonicalize plans with minimal changes.
+     *
+     * We remove any unimportant information like source location, make the variable
+     * names consistent, and order them.
+     *
+     * This is used in context of fragment result caching
+     */
+    DEFAULT,
+    /**
+     * CONNECTOR strategy is used to additionally canoncialize `TableScanNode`s
+     * by giving a connector specific implementation.
+     *
+     * Connector is free to canonicalize table layout and constraints that are pushed to
+     * the TableScanNode. For example, Hive connector will remove constants for the
+     * column values if the column in a partition key.
+     *
+     * This is used in context of history based optimizations.
+     */
+    CONNECTOR,
+    /**
+     * REMOVE_CONSTANTS strategy is used to canonicalize plan and
+     * remove all constants as well.
+     *
+     * This is used in context of history based optimizations.
+     */
+    REMOVE_CONSTANTS,
```

Comment:
Actually have a question for this: is it safe to remove constants? We would plan to be very different if they are
- `select * from A join B on A.col1 = B.col1 where B.col2 between 1 and 1000`
- `select * from A join B on A.col1 = B.col1 where B.col2 between 1 and 10`

The previous query could be a hash join because the build side is huge. The latter could be a broadcast join because the build side is small.